### DISCUSSION TITLE
Building libosdp for windows.

### DISCUSSION THREAD

**Discussion by azar57** [goToMain/libosdp#190] [Q&A] [answered]
Hello!

Can you advise how to build libosdp for windows?
The library built for ubuntu works fine! Installed via pip under windows works fine too! But I can't imagine how to build libosdp for windows, especially the utils submodule. There are a lot of calls to linux system libraries. I can't imagine how to replace all the sys/fcntl.h, sys/socket.h, sys/ipc.h.
I'm wondering how it works with python under windows? As I understand it, python APIs are mostly wrappers over the C library.

**Comment by sidcha** [comment]
I assume you are targeting MSVC; in that case you should be able to compile LibOSDP core without any issues. There are some parts of c-utils that are by design incompatible with Windows but those are not needed to build the library (follow the `—lib-only` flag in configure.sh). 

The python/rust bindings are building LibOSDP for windows natively first before producing the wrappers. You could also consume the library from vcpkg upstream if you are familiar with it.

If all of those combinations don’t work for your use case, my suggestion would be to give it a try and post issues you encounter.

**Comment by azar57** [comment]
Thanks for the reply!

Unfortunately I still can't build libosdp for windows.

Maybe there is something I don't understand about how 'configure -> make -> make install' should work on windows. To run the configure script on windows, the power shell suggested me to install wsl, which I did. Running configure under wsl resulted in a config.make. Then there is a dead end. The resulting Makefile with config.make included is assembled in WSL into a .a file that can be used in WSL, but I don't think I can use it to build a Windows application. 

Installing libosdp with vcpkg was successful. But when I run cmake I get an error like "By not providing "Findlibosdp.cmake" in CMAKE_MODULE_PATH..." it looks like libosdpConfig.cmake or libosdp-config.cmake file is required to use the library.

I tried to build a project with libosdp sources. But it still ends up requiring utils/disjoint_set.c to be built, and that requires utils/utils

**Reply by sidcha** [reply]
> Installing libosdp with vcpkg was successful. But when I run cmake I get an error like "By not providing "Findlibosdp.cmake" in CMAKE_MODULE_PATH..." it looks like libosdpConfig.cmake or libosdp-config.cmake file is required to use the library.

I guess we are missing some documentation about vcpkg builds on Windows. I will give it a try and add it.

> I tried to build a project with libosdp sources. But it still ends up requiring utils/disjoint_set.c to be built, and that requires utils/utils

That is not an issue - not all files in utils/ are windows incompatible; you should be able to build everything that the core library (src/*.c) depends on for windows.

**Reply by azar57** [reply]
> That is not an issue - not all files in utils/ are windows incompatible; you should be able to build everything that the core library (src/*.c) depends on for windows.

I feel like I'm missing something important. To build `osdp_cp.c` you need to build `utils/disjoint_set.c` which requires utils. But how do I build _libosdp_ without utils?

**Reply by sidcha** [reply, marked as answer]
Basically you don't need to build `libutils.a` (the whole library) to build LibOSDP. You can just build the subset of the c-utils project that  `src/*.c` depends on. This is already how LibOSDP is building c-utils (cmake, make, python and rust). Sorry if that wasn't conveyed correctly.

Nevertheless, I made some more fixes now that makes Windows platforms a first class citizen. I also updated the [docs](https://libosdp.sidcha.dev/libosdp/build-and-install#cmake-build) to reflect how you can build on Windows from sources and also added a note on [how to consume LibOSDP from vcpkg](https://libosdp.sidcha.dev/libosdp/build-and-install#using-vcpkg). Let me know if you face any further issues.
